src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -85,29 +86,69 @@ def calculate_timeout_seconds(
     ),
   )
 
-  test_target_sources_file_names = sorted(source_root_stripped_test_target_sources.snapshot.files)
+  return TestTargetSetup(
+    requirements_pex=resolved_requirements_pex,
+    sources_file_names=tuple(sorted(source_root_stripped_test_target_sources.snapshot.files)),
+    input_files_digest=merged_input_files
+  )
+
+
+@rule(name="Run pytest")
+async def run_python_test(
+  test_target: PythonTestsAdaptor,
+  pytest: PyTest,
+  python_setup: PythonSetup,
+  subprocess_encoding_environment: SubprocessEncodingEnvironment
+) -> TestResult:
+  """Runs pytest for one target."""
+
+  test_setup = await Get[TestTargetSetup](PythonTestsAdaptor, test_target)
+
   timeout_seconds = calculate_timeout_seconds(
     timeouts_enabled=pytest.options.timeouts,
     target_timeout=getattr(test_target, 'timeout', None),
     timeout_default=pytest.options.timeout_default,
     timeout_maximum=pytest.options.timeout_maximum,
   )
-  request = resolved_requirements_pex.create_execute_request(
+  request = test_setup.requirements_pex.create_execute_request(
     python_setup=python_setup,
     subprocess_encoding_environment=subprocess_encoding_environment,
-    pex_path=f'./{output_pytest_requirements_pex_filename}',
-    pex_args=(*pytest.options.args, *test_target_sources_file_names),
-    input_files=merged_input_files,
+    pex_path=f'./{test_setup.requirements_pex.output_filename}',
+    pex_args=(*pytest.options.args, *test_setup.sources_file_names),
+    input_files=test_setup.input_files_digest,
     description=f'Run Pytest for {test_target.address.reference()}',
     timeout_seconds=timeout_seconds if timeout_seconds is not None else 9999
   )
   result = await Get[FallibleExecuteProcessResult](ExecuteProcessRequest, request)
   return TestResult.from_fallible_execute_process_result(result)
 
 
+@rule(name="Run pytest in an interactive process")
+async def debug_python_test(
+  test_target: PythonTestsAdaptor,
+  pytest: PyTest,
+  python_setup: PythonSetup,
+  runner: InteractiveRunner
+) -> TestDebugResult:
+
+  test_setup = await Get[TestTargetSetup](PythonTestsAdaptor, test_target)
+  pex_args = (*pytest.options.args, *test_setup.sources_file_names)
```

Comment:
It looks like these are the same between the two rules. Consider adding it as part of `TestTargetSetup` as a field called `args`. Then, this rule would no longer need `pytest` in its signature.